Reworking performance page

sfshaza2 opened the pull request:
As it says.

Staged: https://sz-flutter.firebaseapp.com/perf/best-practices
(Images don't work on staged sites...)

cc @goderbauer @dnfield @kenzieschmoll @johnpryan

kenzieschmoll commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+  you can precalculate what this overlapped,
+  semi-transparent object looks like, cache it,
+  and use that instead of calling `saveLayer()`. 
+  This works with any static shape that you can
+  precalculate.
```
@johnpryan @InMatrix this would be a great place to link to an example. Perhaps a new addition to [perf_dignosis_demo](https://github.com/InMatrix/perf_diagnosis_demo)?

InMatrix replied:
It might be useful to look at where and how `saveLayer` is used in pub packages using [survayor](https://github.com/pq/surveyor). It's possible that Flutter lacks some high-level APIs which pushed users towards `saveLayer`. Cc: @pq

pq replied:
> It might be useful to look at where and how saveLayer is used in pub packages using [survayor](https://github.com/pq/surveyor).

That could be edifying for sure.  Happy to advise if anyone decides to look closer.

kenzieschmoll commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+
+If your code performs unnecessary layout passes,
+it can slow performance. To determine if this
+is your issue, enable the **Track layouts option**
```
link to Track Layouts docs: https://docs.flutter.dev/development/tools/devtools/performance#track-layouts

kenzieschmoll commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+_telling_ each card what size to use. 
+
+If your code performs unnecessary layout passes,
+it can slow performance. To determine if this
```
I think we'll want to tell users to look for the Intrinsics timeline events. Those events were added here: https://github.com/flutter/flutter/pull/93086/files. Looks like the events will have the name '$runtimeType intrinsics', but I have not confirmed yet in DevTools.

sfshaza2 replied:
added a comment to the doc

goderbauer commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+when creating your UI and you might have memorized Flutter's
+basic layout rule: **Constraints go down. Sizes go up.
+Parent sets position.**
+
```
maybe mention somewhere that flutter strives to just do one pass and intrinsics add a second pass.

sfshaza2 replied:
done

goderbauer commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+[PENDING: Where is the track layout option in DevTools?]
+
+[PENDING: How does once determine "excessive"?]
+[PENDING: Is the primary way to fix this by lazy loading
+the grid or list? Or is there something else?]
```
lazy loading wouldn't help for the intrinsics part. You need to determine if you actually need the intrinsic sizing. For the use case described above (sizing children based on the largest size) you can't get around it. If you can determine the size of your children upfront, though, you can avoid intrinsics by just setting the children to a fixed size.

If you want to layout children relative to the size of another child you can also avoid intrinsics by writing a custom render object that lays out that child anchor first and lays out the other children around it. That can be done in a single pass.

sfshaza2 replied:
done

goderbauer commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+for a total of 16ms or less.
+If missing frames (jankiness) is a concern,
+then 16ms for each of
+the build and render stages is OK.
```
I am confused by what this paragraph is trying to tell me. Do I hav 16ms for each of render or build or 16ms for both together?

sfshaza2 replied:
This is an old paragraph. I believe it's saying 16ms for both together. Yes?

sfshaza2 replied:
I deleted the confusing sentence.

goderbauer commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+
+If you need to tune your app’s performance,
+or perhaps the UI isn't as smooth as you expect,
+the Flutter plugin for your IDE can help.
```
@kenzieschmoll Is this also available in devtools?

sfshaza2 replied:
I've added this, but was unsure if I should delete the IDE info. Please check, @kenzieschmoll

domesticmouse reviewed (approved):
Any idea why stable is b0rked?

domesticmouse commented:
Update snippets required by looks of things…

InMatrix commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+and are worth investigating if you are having issues&mdash;you
+might have one of the following scenarios:
+
+* Excessive calls to `savelayer()`.
```
typo `savelayer()` => `saveLayer()`

sfshaza2 replied:
Done

InMatrix commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+  a custom rendering layer.
+
+{{site.alert.secondary}}
+  <h4 markdown="1">When is `savelayer()` required?</h4><a name="when-is-savelayer-required"></a>
```
`savelayer()` => `saveLayer`

khanhnwin commented:
Build was failing on stable due to #7001 - set codelabs submodule to commit [cfdedfb](https://github.com/flutter/codelabs/tree/cfdedfb1cb00fc0cc2a97d42899653b2a9c08dd6) to fix it.

khanhnwin commented:
@sfshaza2 It looks like we're linking to perf/rendering/ui-performance in a few different place and the build is failing since the file's been moved.

sfshaza2 commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+  [`FadeInImage`][] widget, which applies a gradual
+  opacity using the GPU’s fragment shader.
+  For more information, check out the [`Opacity`][] docs.
+* **Clipping** doesn’t call `saveLayer()` (unless
```
Done

sfshaza2 commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,375 @@
+For some widgets, particularly grids and lists,
+the layout process can be expensive.
+For example, consider a 10x500 grid of `Card`s.
+A grid should have uniformly sized cells,
```
done

sfshaza2 commented:
@kenzieschmoll @goderbauer could you take another look? I believe I have addressed feedback.

domesticmouse commented:
Link checker failure: https://github.com/flutter/website/runs/6001365854?check_suite_focus=true#step:3:11279

kenzieschmoll commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,411 @@
+[Working with long lists]: {{site.url}}/cookbook/lists/long-lists
+
+<a href="minimize-layout-passes-caused-by-intrinsic-operations"></a>
+### Minimize layout passes caused by intrinsic operations
```
should this be ## instead of ###? I don't think minimizing layout passes caused by intrinsic operations applies explicitly to Lists and Grids. @goderbauer can you check me on this?

kenzieschmoll commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,411 @@
+write the most performant Flutter app possible.
+
+Sometimes, apps are janky (not smooth) on some platforms.
+We have found two scenarios that can cause jankiness,
```
I worry that saying "We have found two scenarios that can cause jankiness" makes it sound like there are only two reasons your app could be janky, which is not true. Maybe we should not call out saveLayer() and excessive layout passes here at all, but just allow users to read about them below? It seems odd that only these two scenarios get the special treatment of being called out here and in their own section below.

dnfield replied:
Perhaps something like `Some rendering and layout operations are known to be slow, but cannot always be avoided. They should be used thoughtfully, following the guidance below` or something like that?

sfshaza2 replied:
Love it! Done

dnfield commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,411 @@
+  Calling `saveLayer()` allocates an offscreen buffer.
+  Drawing content into the offscreen buffer might trigger
+  render target switches that are particularly slow in
+  older GPUs.
```
I would replace `older` with `mobile`. 

Alternatively, I might express this as 

> The GPU wants to run like a firehose, and a render target switch forces the GPU to redirect that stream temporarily and then direct it back again. On mobile GPUs this is particularly disruptive to rendering throughput.

sfshaza2 replied:
Love it!

dnfield commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,411 @@
+  and use that instead of calling `saveLayer()`. 
+  This works with any static shape that you can precalculate.
+{% comment %}
+TBD: It would be nice if we could link to an example.
+  Kenzie suggested to John and Tao that we add an
+  example to perf_diagnosis_demo.)
```
I think @goderbauer had a nice constructed example for this?

goderbauer replied:
I don't have one for saveLayer, sadly.

dnfield commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,411 @@
+{% endcomment %}
+
+* If the calls are coming from a package that you don't own,
+  contact the package owner and ask why
+  these calls are necessary? Can they be reduced or
+  eliminated? If not, you might need to find another
+  package, or write your own.
```
Applying some pressure to the owner might be good. It would be good to have guidance for owners here, like providing documentation of when a saveLayer might be necessary for their package and how to avoid it/when you cannot avoid it.

sfshaza2 replied:
ok, please double check my edits

dnfield commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,411 @@
+  eliminated? If not, you might need to find another
+  package, or write your own.
+
+* Perhaps you could write a (simpler) [`CustomPainter`][].
+
+* If all else fails, you might consider implementing
+  a custom [`RenderObject`][].
```
using or avoid a saveLayer has less to do with whether you use a CustomPainter or a RenderObject and more to do with how you achieve your visual effects.

For example, if you need to accurately blend multiple overlapping semi-transparent shapes, you must use a saveLayer. But if you can refactor your painting logic to avoid the overlaps, you can avoid the saveLayer.

sfshaza2 replied:
done, but please recheck

sfshaza2 commented:
@dnfield @kenzieschmoll @goderbauer, I want to really thank all of your for your excellent feedback!!!

For this commit, I've:
* Added redirects as the perf/rendering docs have now moved.
* Reoganized the best practices page a bit more. It's staged, once again, [here](https://sz-flutter.firebaseapp.com/perf/best-practices).

Please check!

goderbauer commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,425 @@
+    which uses this principle to avoid rebuilding its
+    descendents when animating.
+
+Use `const` constructors on widgets as much as possible,
```
cross-linking opportunity: If apps have the recommended lints from package:flutter_lints enabled, they will automatically be reminded to use `const` where possible. If an app isn't using the recommended set of lints, they may want to consider upgrading to it as explained in https://docs.flutter.dev/release/breaking-changes/flutter-lints-package#migration-guide.

goderbauer commented on src/perf/best-practices.md:
```diff
@@ -0,0 +1,425 @@
+to learn how many layout passes were performed.
+{% comment %}
+TBD: From Kenzie: Tell users to look for intrinsic timeline events.
+They will possibly be named: '$runtimeType intrinsics'.
```
That is the current name the framework gives these events.

sfshaza2 replied:
thx!

goderbauer reviewed (approved):
LGTM

Thanks for writing this up!

kenzieschmoll reviewed (approved):
thanks Shams!

domesticmouse commented:
Conflicts in need of resolution, and link checker is still upset. Poor CI bot.

khanhnwin reviewed (approved):
LGTM!
